library/auth_queries.c

```diff
@@ -0,0 +1,323 @@
+            mfa_enroll_cb_ctx->cb(mfa_enroll_cb_ctx->ztx, err->err, NULL, mfa_enroll_cb_ctx->cb_ctx);
+        }
+        FREE(err)
+    }
+
+    if (mfa_enrollment == NULL) {
+        ziti_ctrl_post_mfa(&mfa_enroll_cb_ctx->ztx->controller, ziti_mfa_enroll_post_internal_cb, ctx);
+    } else {
+        mfa_enroll_cb_ctx->cb(mfa_enroll_cb_ctx->ztx, ZITI_OK, mfa_enrollment, mfa_enroll_cb_ctx->cb_ctx);
+        FREE(ctx)
+        free_ziti_mfa_enrollment(mfa_enrollment);
+    }
+}
+
+void ziti_mfa_enroll(ziti_context ztx, ziti_mfa_enroll_cb enroll_cb, void *ctx) {
+    NEWP(mfa_enroll_cb_ctx, ziti_mfa_enroll_cb_ctx);
+
+    mfa_enroll_cb_ctx->ztx = ztx;
+    mfa_enroll_cb_ctx->cb = enroll_cb;
+    mfa_enroll_cb_ctx->cb_ctx = ctx;
+
+    ziti_ctrl_get_mfa(&ztx->controller, ziti_mfa_enroll_get_internal_cb, mfa_enroll_cb_ctx);
+}
+
+void ziti_mfa_remove_internal_cb(void *empty, ziti_error *err, void *ctx) {
+    ziti_mfa_cb_ctx *mfa_cb_ctx = ctx;
+
+    if (err != NULL) {
+        ZITI_LOG(ERROR, "error during remove MFA call: %d - %s - %s", err->http_code, err->code, err->message);
+        mfa_cb_ctx->cb(mfa_cb_ctx->ztx, err->err, mfa_cb_ctx->cb_ctx);
+        FREE(err)
+    } else {
+        mfa_cb_ctx->cb(mfa_cb_ctx->ztx, ZITI_OK, mfa_cb_ctx->cb_ctx);
+    }
+
+    FREE(mfa_cb_ctx->code)
+    FREE(ctx)
+}
+
+void ziti_mfa_remove(ziti_context ztx, char *code, ziti_mfa_cb remove_cb, void *ctx) {
+    NEWP(mfa_cb_ctx, ziti_mfa_cb_ctx);
+    mfa_cb_ctx->ztx = ztx;
+    mfa_cb_ctx->cb = remove_cb;
+    mfa_cb_ctx->cb_ctx = ctx;
+    mfa_cb_ctx->code = strdup(code);
+
+    ziti_ctrl_delete_mfa(&ztx->controller, mfa_cb_ctx->code, ziti_mfa_remove_internal_cb, mfa_cb_ctx);
+}
+
+void ziti_mfa_verify_internal_cb(void *empty, ziti_error *err, void *ctx) {
+    ziti_mfa_cb_ctx *mfa_cb_ctx = ctx;
+
+    if (err != NULL) {
+        ZITI_LOG(ERROR, "error during verify MFA call: %d - %s - %s", err->http_code, err->code, err->message);
+        mfa_cb_ctx->cb(mfa_cb_ctx->ztx, err->err, mfa_cb_ctx->cb_ctx);
+        FREE(err)
+    } else {
+        mfa_cb_ctx->cb(mfa_cb_ctx->ztx, ZITI_OK, mfa_cb_ctx->cb_ctx);
+    }
+
+    FREE(ctx)
+}
+
+void ziti_mfa_verify(ziti_context ztx, char *code, ziti_mfa_cb verify_cb, void *ctx) {
+    NEWP(mfa_cb_ctx, ziti_mfa_cb_ctx);
+    mfa_cb_ctx->ztx = ztx;
+    mfa_cb_ctx->cb = verify_cb;
+    mfa_cb_ctx->cb_ctx = ctx;
+
+    char *body = ziti_mfa_code_body(code);
+
+    ziti_ctrl_post_mfa_verify(&ztx->controller, body, strlen(body), ziti_mfa_verify_internal_cb, mfa_cb_ctx);
+}
+
+void ziti_auth_query_mfa_auth_internal_cb(void *empty, ziti_error *err, void *ctx) {
+    ziti_mfa_auth_ctx *mfa_ctx = ctx;
+    ziti_context ztx = mfa_ctx->ztx;
+    if (err != NULL) {
+        ZITI_LOG(ERROR, "error during MFA auth call: %d - %s - %s", err->http_code, err->code, err->message);
+        mfa_ctx->attempts++;
+
+        if (mfa_ctx->attempts >= MFA_MAX_ATTEMPTS) {
+            exit(err->err);
```

Comment:
Wasn't sure what to do if someone just fails over and over to put the right TOTP/recovery code in.

Reply:
![alt text](https://static.miraheze.org/allthetropeswiki/5/5a/Detonator_6550.jpg)

Reply:
this is bad idea when we allow multiple identities

Reply:
I simply removed it. The external application can track attempts and let it decide how it wants to handle the situation (including giving up on the user).